tile_downloader.py

```diff
@@ -1,165 +1,274 @@
-from pygeotile.tile import Tile
-from shapely.geometry import Polygon
+from pyproj import transform, Proj
 
 import maps
-from utils import ImageFormat, get_tile_gen, get_filename, get_bbox_in_tms, get_tiles_bbox
+from _tile_downloader import download_in_gtiff as _download_in_gtiff, download_tiles as _download_tiles, \
+    construct_gtiff as _construct_gtiff
+from utils import ImageFormat
 
-CRS = 'EPSG:4326'
 
+def _get_projection(**kwargs):
+    proj = kwargs.get('projection')
 
-def download_tiles(
-        tiles_dir: Union[Path, str],
-        bbox: Tuple[float, float, float, float],
-        zoom: int,
-        map_: Type[maps.Map],
-        proxies: Optional[dict] = None
-) -> None:
-    # language=rst
-    """
-    Download tiles from `bbox` with `zoom` zoomlevel to `tiles_dir` from `map_` using `proxies`.
-    :param tiles_dir:
-    :param bbox: area of the geo coordinates in the form of:
-     `(min_lat, min_lon, max_lat, max_lon)`
-    :param zoom:
-    :param map_:
-    :param proxies:
-    :return:
-    """
-    session = requests.session()
-    if proxies is not None:
-        session.proxies = proxies
+    if 'crs' in kwargs:
+        if proj is not None:
+            raise TypeError
 
-    for tile in get_tile_gen(bbox, zoom):
-        path = get_filename(tile, map_.tiles_format, tiles_dir)
+        proj = Proj(kwargs['crs'])
 
-        if not path.exists():
-            for url in map_.get_urls_gen(tile):
-                response = session.get(url)
-                time.sleep(map_.get_timeout())
-                if response.ok:
-                    with open(path, 'wb') as file:
-                        file.write(response.content)
-                    break
+    if 'srs' in kwargs:
+        if proj is not None:
+            raise TypeError
 
-    session.close()
+        proj = Proj(kwargs['srs'])
 
+    return Proj(init='EPSG:4326') if proj is None else proj
 
-def _merge_tiles(
-        path: Union[Path, str],
+
+def _get_area_args_as_bbox(**kwargs):
+    bbox = kwargs.get('bbox')
+
+    if {'min_x', 'min_y', 'max_x', 'max_y'} & kwargs.keys():
+        if bbox is not None:
+            raise TypeError
+
+        bbox = kwargs['min_x'], kwargs['min_y'], kwargs['max_x'], kwargs['max_y']
+
+    if {'left', 'right', 'bottom', 'top'} & kwargs.keys():
+        if bbox is not None:
+            raise TypeError
+
+        bbox = kwargs['left'], kwargs['bottom'], kwargs['right'], kwargs['top']
+
+    if {'min_lat', 'min_lon', 'max_lat', 'max_lon'} & kwargs.keys():
+        if bbox is not None:
+            raise TypeError
+
+        proj = _get_projection(**kwargs)
+        bbox = (
+            transform(Proj(proj='latlong'), proj, kwargs['min_lon'], kwargs['min_lat']) +
+            transform(Proj(proj='latlong'), proj, kwargs['max_lon'], kwargs['max_lat'])
+        )
+
+    if bbox is None:
+        raise TypeError
+
+    return bbox
+
+
+def _get_zoom(**kwargs):
+    zoom = kwargs.get('zoom')
+
+    if 'zoomlevel' in kwargs:
```

Comment:
You can get "zoom" and "zoomlevel" both in kwargs.get() method. And then check that only one can has None-value.

Reply:
Also you can check zoom datatype and that it is above zero